File: impl-large/csharp/Domain.cs
// Task Service, large tier — domain types, constants and pure rules.

using System.Text.Json;

record Detail(string Field, string Message);

record Session(string Token, int UserId)
{
    public int Used { get; set; }
}

record User(int Id, string Username, string Password)
{
    public required string Role { get; set; }
    public required int Quota { get; set; }
    public int Version { get; set; } = 1;
    public bool Deleted { get; set; }
}

record Project(int Id, int OwnerId)
{
    public required string Name { get; set; }
    public int Version { get; set; } = 1;
    public bool Deleted { get; set; }
}

record TaskItem(int Id, int ProjectId)
{
    public required string Title { get; set; }
    public required int Priority { get; set; }
    public required string Status { get; set; }
    public required int? AssigneeId { get; set; }
    public required string InternalNote { get; set; }
    public int Version { get; set; } = 1;
    public bool Deleted { get; set; }
}

record Comment(int Id, int TaskId, int AuthorId, string Body);

record AuditEntry(int Seq, int ActorId, string Action, string Resource, int ResourceId);

record OutboxEvent(int Seq, string Name, int ResourceId)
{
    public bool Delivered { get; set; }
}

/// <summary>Every failure path throws this. The api layer turns it into the envelope.</summary>
class AppError(int status, string code, string message, List<Detail>? details = null)
    : Exception(message)
{
    public int Status { get; } = status;
    public string Code { get; } = code;
    public List<Detail> Details { get; } = details ?? [];
}

static class Domain
{
    public const int MaxTitleLength = 80;
    public const int MaxNameLength = 60;
    public const int MaxCommentLength = 200;
    public const int MaxBulkItems = 20;
    public const int MinPriority = 1;
    public const int MaxPriority = 5;
    public const int DefaultLimit = 20;
    public const int MaxLimit = 100;
    public const int DefaultQuota = 10000;
    public const int ProbeQuota = 5;
    public const int Port = 8080;

    public static readonly string[] Roles = ["admin", "user"];
    public static readonly string[] Statuses = ["todo", "in_progress", "done", "archived"];
    public static readonly string[] ProjectSorts = ["id", "name", "taskCount"];
    public static readonly string[] TaskSorts = ["id", "title", "priority", "score", "status"];
    public static readonly string[] UserSorts = ["id", "username", "role"];
    public static readonly string[] CommentSorts = ["id", "authorId"];
    public static readonly string[] SeqSorts = ["seq"];
    public static readonly string[] GroupBys = ["assignee", "status", "project"];

    public static readonly Dictionary<string, int> StatusBonus = new()
    {
        ["todo"] = 0, ["in_progress"] = 3, ["done"] = 5, ["archived"] = 0,
    };

    public static readonly HashSet<(string, string)> Transitions =
    [
        ("todo", "in_progress"), ("todo", "archived"), ("in_progress", "todo"),
        ("in_progress", "done"), ("done", "archived"),
    ];

    public static AppError BadRequest() => new(400, "bad_request", "the request is malformed");

    public static AppError Unauthorized() => new(401, "unauthorized", "authentication is required");

    public static AppError InvalidCredentials() =>
        new(401, "invalid_credentials", "the username or password is wrong");

    public static AppError Forbidden() => new(403, "forbidden", "you may not access this resource");

    public static AppError NotFound() => new(404, "not_found", "the resource does not exist");

    public static AppError Conflict() => new(409, "conflict", "the resource already exists");

    public static AppError InvalidTransition() =>
        new(409, "invalid_transition", "the status change is not allowed");

    public static AppError PreconditionFailed() =>
        new(412, "precondition_failed", "the resource has changed");

    public static AppError PreconditionRequired() =>
        new(428, "precondition_required", "the If-Match header is required");

    public static AppError QuotaExceeded() =>
        new(429, "quota_exceeded", "the request quota is exhausted");

    public static AppError Invalid(List<Detail> details) =>
        new(422, "validation_failed", "the request body is not valid",
            [.. details.OrderBy(entry => entry.Field, StringComparer.Ordinal)
                .ThenBy(entry => entry.Message, StringComparer.Ordinal)]);

    public static Detail Fail(string field, string message) => new(field, message);

    public static int ComputeScore(int priority, string status)
    {
        var baseScore = priority * 10;
        return baseScore + StatusBonus[status];
    }

    public static void CheckString(string value, string fieldName, int maxLength, List<Detail> errors)
    {
        if (value.Length == 0) errors.Add(Fail(fieldName, $"{fieldName} is required"));
        else if (value.Length > maxLength) errors.Add(Fail(fieldName, $"{fieldName} is too long"));
    }

    public static void CheckPriority(int? value, List<Detail> errors)
    {
        if (value is null || value < MinPriority || value > MaxPriority)
            errors.Add(Fail("priority", "priority is out of range"));
    }

    public static void CheckStatus(string? value, List<Detail> errors)
    {
        if (value is null || !Statuses.Contains(value))
            errors.Add(Fail("status", "status is not valid"));
    }

    public static void CheckRole(string? value, List<Detail> errors)
    {
        if (value is null || !Roles.Contains(value)) errors.Add(Fail("role", "role is not valid"));
    }

    public static void CheckQuota(int? value, List<Detail> errors)
    {
        if (value is null || value < 0) errors.Add(Fail("quota", "quota is out of range"));
    }

    public static JsonElement? Node(JsonElement body, string field) =>
        body.TryGetProperty(field, out var node) ? node : null;

    /// <summary>A string field. A present value of another type is malformed.</summary>
    public static string Text(JsonElement body, string field, string fallback = "")
    {
        if (!body.TryGetProperty(field, out var node)) return fallback;
        if (node.ValueKind != JsonValueKind.String) throw BadRequest();
        return node.GetString() ?? "";
    }

    /// <summary>An integer field. A present value of another type is malformed.</summary>
    public static int? Whole(JsonElement body, string field, int? fallback)
    {
        if (!body.TryGetProperty(field, out var node)) return fallback;
        if (node.ValueKind == JsonValueKind.Null) return null;
        if (node.ValueKind == JsonValueKind.Number && node.TryGetInt32(out var value)) return value;
        throw BadRequest();
    }

    /// <summary>A string field whose wrong type is a validation problem, not a parse error.</summary>
    public static string? LooseText(JsonElement body, string field, string? fallback)
    {
        if (!body.TryGetProperty(field, out var node)) return fallback;
        return node.ValueKind == JsonValueKind.String ? node.GetString() : null;
    }

    /// <summary>An integer field whose wrong type is a validation problem, not a parse error.</summary>
    public static int? LooseWhole(JsonElement body, string field, int? fallback)
    {
        if (!body.TryGetProperty(field, out var node)) return fallback;
        return node.ValueKind == JsonValueKind.Number && node.TryGetInt32(out var value)
            ? value : null;
    }
}

File: impl-large/typescript/src/domain.ts
// Task Service, large tier — domain types, constants and pure rules.

export const MAX_TITLE_LENGTH = 80;
export const MAX_NAME_LENGTH = 60;
export const MAX_COMMENT_LENGTH = 200;
export const MAX_BULK_ITEMS = 20;
export const MIN_PRIORITY = 1;
export const MAX_PRIORITY = 5;
export const DEFAULT_LIMIT = 20;
export const MAX_LIMIT = 100;
export const DEFAULT_QUOTA = 10000;
export const PROBE_QUOTA = 5;
export const PORT = 8080;

export const ROLES = ["admin", "user"];
export const STATUSES = ["todo", "in_progress", "done", "archived"];
export const STATUS_BONUS: Record<string, number> = {
  todo: 0,
  in_progress: 3,
  done: 5,
  archived: 0,
};
export const TRANSITIONS = new Set([
  "todo->in_progress",
  "todo->archived",
  "in_progress->todo",
  "in_progress->done",
  "done->archived",
]);
export const ACTIONS = ["create", "update", "delete", "restore"];
export const PROJECT_SORTS = ["id", "name", "taskCount"];
export const TASK_SORTS = ["id", "title", "priority", "score", "status"];
export const USER_SORTS = ["id", "username", "role"];
export const COMMENT_SORTS = ["id", "authorId"];
export const SEQ_SORTS = ["seq"];
export const GROUP_BYS = ["assignee", "status", "project"];

export interface User {
  id: number;
  username: string;
  password: string;
  role: string;
  quota: number;
  version: number;
  deleted: boolean;
}

export interface Session {
  token: string;
  userId: number;
  used: number;
}

export interface Project {
  id: number;
  name: string;
  ownerId: number;
  version: number;
  deleted: boolean;
}

export interface Task {
  id: number;
  projectId: number;
  title: string;
  priority: number;
  status: string;
  assigneeId: number | null;
  internalNote: string;
  version: number;
  deleted: boolean;
}

export interface Comment {
  id: number;
  taskId: number;
  authorId: number;
  body: string;
}

export interface AuditEntry {
  seq: number;
  actorId: number;
  action: string;
  resource: string;
  resourceId: number;
}

export interface OutboxEvent {
  seq: number;
  name: string;
  resourceId: number;
  delivered: boolean;
}

export interface Detail {
  field: string;
  message: string;
}

export type Body = Record<string, unknown>;
export type Row = Record<string, unknown>;

/** Every failure path throws this. The api layer turns it into the envelope. */
export class AppError {
  constructor(
    readonly status: number,
    readonly code: string,
    readonly message: string,
    readonly details: Detail[] = [],
  ) {}
}

export function badRequest(): AppError {
  return new AppError(400, "bad_request", "the request is malformed");
}

export function unauthorized(): AppError {
  return new AppError(401, "unauthorized", "authentication is required");
}

export function invalidCredentials(): AppError {
  return new AppError(401, "invalid_credentials", "the username or password is wrong");
}

export function forbidden(): AppError {
  return new AppError(403, "forbidden", "you may not access this resource");
}

export function notFound(): AppError {
  return new AppError(404, "not_found", "the resource does not exist");
}

export function conflict(): AppError {
  return new AppError(409, "conflict", "the resource already exists");
}

export function invalidTransition(): AppError {
  return new AppError(409, "invalid_transition", "the status change is not allowed");
}

export function preconditionFailed(): AppError {
  return new AppError(412, "precondition_failed", "the resource has changed");
}

export function preconditionRequired(): AppError {
  return new AppError(428, "precondition_required", "the If-Match header is required");
}

export function quotaExceeded(): AppError {
  return new AppError(429, "quota_exceeded", "the request quota is exhausted");
}

export function invalid(details: Detail[]): AppError {
  details.sort((a, b) =>
    a.field < b.field ? -1
      : a.field > b.field ? 1
      : a.message < b.message ? -1
      : a.message > b.message ? 1
      : 0,
  );
  return new AppError(422, "validation_failed", "the request body is not valid", details);
}

export function fail(field: string, message: string): Detail {
  return { field, message };
}

export function computeScore(priority: number, status: string): number {
  const baseScore = priority * 10;
  return baseScore + STATUS_BONUS[status];
}

export function checkString(
  value: string,
  fieldName: string,
  maxLength: number,
  errors: Detail[],
): void {
  if (value === "") {
    errors.push(fail(fieldName, `${fieldName} is required`));
  } else if (value.length > maxLength) {
    errors.push(fail(fieldName, `${fieldName} is too long`));
  }
}

export function checkPriority(value: number | null, errors: Detail[]): void {
  if (value === null || value < MIN_PRIORITY || value > MAX_PRIORITY) {
    errors.push(fail("priority", "priority is out of range"));
  }
}

export function checkStatus(value: unknown, errors: Detail[]): void {
  if (typeof value !== "string" || !STATUSES.includes(value)) {
    errors.push(fail("status", "status is not valid"));
  }
}

export function checkRole(value: unknown, errors: Detail[]): void {
  if (typeof value !== "string" || !ROLES.includes(value)) {
    errors.push(fail("role", "role is not valid"));
  }
}

export function checkQuota(value: unknown, errors: Detail[]): void {
  if (typeof value !== "number" || !Number.isInteger(value) || value < 0) {
    errors.push(fail("quota", "quota is out of range"));
  }
}
